sycl/source/backend.cpp

```diff
@@ -170,8 +179,12 @@ make_kernel_bundle(pi_native_handle NativeHandle, const context &TargetContext,
   pi::PiProgram PiProgram = nullptr;
   Plugin->call<PiApiKind::piextProgramCreateWithNativeHandle>(
       NativeHandle, ContextImpl->getHandleRef(), !KeepOwnership, &PiProgram);
-  if (ContextImpl->getBackend() == backend::opencl)
-    Plugin->call<PiApiKind::piProgramRetain>(PiProgram);
+  if (ContextImpl->getBackend() == backend::opencl) {
+    if (clRetainProgram(pi::cast<cl_program>(NativeHandle)) != CL_SUCCESS) {
+      throw runtime_error("Failed to retain opencl program",
+                          PI_ERROR_INVALID_PROGRAM);
+    }
+  }
```

Comment:
Handles were casted directly before so calling `piProgramRetain` was the same as `clRetainProgram` but now as there is a UR layer for the handles that is not true anymore, so we need to retain this manually with opencl entry.